### PR TITLE
Bugfix/list of dicts support

### PR DESCRIPTION
**Pull request by austinwiltshire**
Passing in a list of dicts for the parameter space resulted in cv_results_ not being able to be converted to a dataframe. This was due to some parameter fit runs not always using all parameters, such as the case with sklearn's SVM's different kernels.

Using list.extend only actually added things to a list in the cv_results dictionary when the field was there, and the lists often got out of sync with each other, making the resulting dictionary of lists unable to be consumed by pandas.DataFrame

This uses the approach found in sklearn.gridsearch of masked arrays, and keeping track of the index we're actively filling in the resulting dictionary of masked arrays to keep everything aligned.

### CODE REVIEW

File: skopt/searchcv.py [hunk 1]
```diff
@@ -638,13 +652,34 @@ def fit(self, X, y=None, groups=None, callback=None):
...
-        self.cv_results_ = defaultdict(list)
+        n_candidates = len(search_spaces) * self.n_iter
+
+        # We use a masked array as the type of list to store our results in
+        # since not all params are guaranteed to show up in every search -
+        # especially if we take in a list of dicts approach
+        self.cv_results_ = defaultdict(
+            partial(
+                MaskedArray,
+                np.empty(n_candidates,),
```
Comment (CoitThomas) [inline skopt/searchcv.py]: this syntax looks new and scary to me

**Review by CoitThomas — APPROVED** [review]
Looks good to me. Is there a unit test related to what you've refactored to verify the desired results?

**Comment by austinwiltshire** [comment]
Unfortunately scikit optimize isn't being maintained and their unit tests don't appear to run.

I'll be exercising the code inside ensemble though in tests there, is that good enough @CoitThomas ?

**Comment by CoitThomas** [comment]
Yeah, if ensemble tests are able to verify then that's good enough for me 👍
